typing/printtyp.ml

```diff
@@ -1316,12 +1316,38 @@ let prepare_type_constructor_arguments = function
   | Cstr_tuple l -> List.iter prepare_type l
   | Cstr_record l -> List.iter (fun l -> prepare_type l.ld_type) l
 
-let rec tree_of_type_decl id decl =
+let tree_of_label l =
+  (Ident.name l.ld_id, l.ld_mutable = Mutable, tree_of_typexp Type l.ld_type)
 
-  reset_except_context();
+let tree_of_constructor_arguments = function
+  | Cstr_tuple l -> tree_of_typlist Type l
+  | Cstr_record l -> [ Otyp_record (List.map tree_of_label l) ]
 
-  let params = filter_params decl.type_params in
+let tree_of_constructor cd =
+  let name = Ident.name cd.cd_id in
+  let ret = Option.map (tree_of_typexp Type) cd.cd_res in
+  let args = tree_of_constructor_arguments cd.cd_args in
+  {
+      ocstr_name = name;
+      ocstr_args = args;
+      ocstr_return_type = ret;
+  }
 
+(* When printing GADT constructor, we need to forget the naming decision we took
+  for the type parameters and constraints. Indeed, in
+  {[
+  type 'a t = X: 'a -> 'b t
+   ]}
+  It is fine to print both the type parameter ['a] and the existentially
+  quantified ['a] in the definition of the constructor X as ['a]
+ *)
+let isolated_tree_of_constructor cd =
+  match cd.cd_res with
+  | None -> tree_of_constructor cd
+  | Some _ -> Names.with_local_names (fun () -> tree_of_constructor cd)
```

Comment:
hmm, not fond of this naming structure where the obvious function name `tree_of_constructor` is silently doing the wrong thing in some case and all callers should instead call the non-obvious variant `isolated_tree_of_constructor`. Maybe `tree_of_constructor_aux` and then `tree_of_constructor`? (I'm also not fond of using the same name twice with shadowing.)

Reply:
There is exactly one caller of the `isolated` version: the printer for type declaration, which needs to navigate between the scope shared with the type parameters and an isolated one for GADTs:
```ocaml
type 'a t  (* shared scope *) =
  | (* shared scope *) A
  | (* shared scope *) B of 'a 
  | (* isolated scope *) X: 'a ->  'b t
  | (* isolated scope *) Y: 'a ->  'b * 'b t  
  | (* return to the shared scope *) C of 'a * 'a t
  ```
Outside of this specific context, the `tree_of_constructor` is the correct function to call.